LLaMA 2-7B Components

arifiorino opened the pull request:
- New struct `SRS` which has X and Y elements in G1 and G2, and in both affine and projective form
- `run`, `setup`, `prove`, `verify` all operate on a vector of commitments
- `setup` and `prove` have affine points as input and projective points as output
- New BasicBlocks:
  - `CQ2`: Batched CQ which takes in two tables and two input vectors
  - `MatMul`: Takes row commitments of `A` and column commitments of `B` and outputs row commitments of `C=AB`
  - `MulScalar` and `MulConst` because multiplying by an unknown scalar or a known constant is much faster than a full multiplication argument
  - `Add`/`Subtract` now allow adding a scalar to a vector
  - `Transpose`: Transposes a matrix
  - `Sum`: Sums all the values in a matrix
  - `Concat`: Concatenates the rows of a matrix into a single vector
  - `Alternating`: Checks that a vector `A` alternating with a vector `B` equals a vector `C`
  - `Eq`: Checks that two vectors are equal
- Other new BasicBlocks proved via `CQ2`:
  - `DivConst`, `Exp`, `Log`, `Sqrt`
- New basic blocks proved by other means:
  - `Combine` combines two vectors into a single one by alternating between the two
  - `Split` splits a vector in two by taking odd and even indices
  - `DivScalar` divides a vector by a scalar and produces both the quotient and remainder vectors
  - `RoPE` returns the constants used in the RoPE operation
- `CQ` and `CQ2` save their internal `HashMap` so it doesn't have to be recalculated in every proof
- Created tests in `tests.rs` for all new BasicBlocks that need them

**New 4/1/2024:**
- Added `BatchedBasicBlock` which runs on `NDArray`
- Switched `Graph` to run on `BatchedBasicBlock`
- `BatchedBasicBlock` checks the dimensions of the `model` and `input` before `run`/`setup`/`prove`/`verify` methods

arifiorino committed:
Initial matmul

arifiorino committed:
Added AProof and BProof structs and comments

arifiorino committed:
Added blinding and unit tests

arifiorino committed:
Faster verification

arifiorino committed:
CQ table caching, vector inputs

arifiorino committed:
BasicBlock multiple outputs

arifiorino committed:
Matrix by matrix multiplication

arifiorino committed:
SRS struct and moving setup/verify to Affine

arifiorino committed:
Unit tests

arifiorino committed:
Add transpose

arifiorino committed:
Changing model to vector of commitments

arifiorino committed:
CQ2

arifiorino committed:
Multiple input nodes

arifiorino committed:
Add Softmax helpers

arifiorino committed:
Fr bug and cq ranges

arifiorino committed:
Add RMSNorm helpers

arifiorino committed:
Add and subtract scalar support and eq

arifiorino committed:
formatting, div error, graph prev, and gen_cq_table

arifiorino committed:
Add alternating argument for RoPE

arifiorino committed:
Finalizing rope

arifiorino committed:
Changed graph

arifiorino committed:
Add concat argument

arifiorino committed:
Add back cqlin and formatting

ddkang commented:
This PR is quite difficult to review, let's try to break things up in the future

Vimalakirti commented:
Haven't read all the codes yet, but here's one quick review:
I notice that we no longer use "ArrayD" after this PR since we perform row-wise proving for all operations now; however, it also means we need to do additional works to handle shape info when creating graph.

ddkang commented on src/graph.rs:
```diff
@@ -1,81 +1,57 @@
-        }
-      }
-    }
+  pub fn run(&self, inputs: &Vec<&Vec<Fr>>, models: &Vec<&Vec<&Vec<Fr>>>) -> Vec<Vec<Vec<Fr>>> {
```
This seems really ugly, I don't understand why we aren't using ArrayD

ddkang replied:
As I've mentioned several times, this prevents us from enforcing any shape constraints on the vectors

ddkang replied:
If this is externally called (via another library) this can lead to unexpected or unsafe behavior

arifiorino replied:
For now we are using ArrayD in `BatchedBasicBlock` which will eventually be changed to a `Layer`

Vimalakirti commented on src/basic_block/concat.rs:
```diff
@@ -0,0 +1,172 @@
+use ark_std::{ops::Mul, ops::Sub, One, UniformRand, Zero};
+use rand::{rngs::StdRng, SeedableRng};
+
+fn calc_pow(alpha: Fr, n: usize) -> Vec<Fr> {
```
This function duplicates everywhere. I suggest we may move it to `util.rs`.

arifiorino replied:
The functions are not exactly the same, note that the one in `MatMul` is not the same as the one in `Alternating`.

Vimalakirti commented on src/basic_block/eq.rs:
```diff
@@ -0,0 +1,68 @@
+    assert!(lhs == rhs);
+  }
+}
+pub struct AddBasicBlock;
```
Why do we need this? We already have `AddBasicBlock` (same struct name) in `add.rs`.

arifiorino replied:
Yes, this was a mistake, I deleted it.

Vimalakirti commented on src/basic_block/matmul.rs:
```diff
@@ -0,0 +1,198 @@
+    for i in 0..l {
+      for j in 0..n {
+        for k in 0..m {
+          r[i][j] += inputs[i][k] * inputs[l + j][k];
```
I know `l` is the boundary to separate elements in A and those in B for matmult A*B. However, I personally find it hard to maintain the code. Do we have better way(s) to realize this?

arifiorino replied:
The three parameters to matmul are `l`,`m`,`n`. We have that `A` is `lxm`, `B` is `mxn`, and `C` is `lxn`. (We assume `AB=C`).
The inputs to `MatMulBasicBlock` are the row commitments to `A` (there are `l` of them), and the column commitments to `B` (there are `n` of them). We don't need to have all `l`,`m`,`n` as parameters of `MatMulBasicBlock`, we only need `l` because `m` and `n` can be inferred using `l`.

liliatangxy replied:
Is there any performance benefit to combining the A and B row commitments vs providing them as separate inputs? Although it isn't difficult to figure out, it hurts readability somewhat

Vimalakirti commented on src/basic_block/max.rs:
```diff
@@ -0,0 +1,18 @@
+    let mut max = Fr::zero();
+    for i in 0..inputs.len() {
+      for j in 0..inputs[0].len() {
+        if inputs[i][j] < Fr::from(1 << 28) && inputs[i][j] > max {
```
I know `1<<28` is the boundary for us to separate positive and negative numbers. I suggest we can make it a const and put it in `util.rs`

arifiorino replied:
For now this is unnecessary.

Vimalakirti commented on src/tests.rs:
```diff
@@ -2,61 +2,92 @@ use crate::basic_block::*;
 }
 
 #[test]
 fn testBasicBlocks() {
```
I've run everything here and nothing went wrong. However, I suggest we can start to think about how to do benchmark here or under `bench/`

Vimalakirti replied:
Also, I think it is good to showcase common ML operations here (like `ReLUBasicBlock`, `RoPEBasicBlock`, and `DivScalarBasicBlock`

liliatangxy replied:
Probably for any block that doesn't have its own prove/verify block.

arifiorino replied:
Right now all the basic blocks that have a prove/verify function are tested in this file.
The other blocks don't need to be tested because they don't require any proof/verification.

Vimalakirti commented on src/basic_block/alternating.rs:
```diff
@@ -0,0 +1,162 @@
+    let beta_pow_g2: G2Affine = util::msm::<G2Projective>(&srs.X2A, &beta_pow_coeff).into();
+    let alpha_beta_pow: Vec<Fr> = (0..2 * n).map(|i| if i % 2 == 0 { alpha_pow[i / 2] } else { beta_pow[i / 2] }).collect();
+    let alpha_beta_pow_coeff = domain_2n.ifft(&alpha_beta_pow);
+    let alpha_beta_pow_g2: G2Affine = util::msm::<G2Projective>(&srs.X2A, &alpha_beta_pow_coeff).into();
```
These msms are really expensive for verifier, but fortunately, they can precompute these. Maybe we can design a setup function for verifier (?)

liliatangxy replied:
Wouldn't we still have to count them in the timing? 
We also may be able to use the same alpha and beta across blocks, so we wouldn't have to redo this computation for each block.

arifiorino replied:
Yes, in the future we will repeat the same alpha and beta across blocks and precompute these MSMs.

Vimalakirti commented on src/graph.rs:
```diff
@@ -1,81 +1,57 @@
   pub basic_block: usize,
-  pub input_nodes: Vec<usize>,
-  pub output_nodes: Vec<usize>,
+  pub inputs: Vec<(i32, usize)>, //(node, output #)
```
It may be good to put `models` (i.e., the weight) as one of the attributes for the corresponding Node (?)

arifiorino replied:
No, this is not a good idea. This is because the model is potentially repeated across multiple nodes. For example there is one model for the cq relu table that only needs to be defined once.

Vimalakirti commented on src/main.rs:
```diff
@@ -14,89 +15,78 @@ mod tests;
+  let mut inputs = vec![&input, &input2];
+  inputs.append(&mut matrix);
+  let empty = vec![];
+  let (id, relu_cq_table) = util::gen_cq_table(vec![&graph.basic_blocks[1]], -(1 << 5), 1 << 6);
```
We need this after the `BasicBlock` of every nonlinear function (sigmoid, tanh). It might be good to write a helper function later

arifiorino replied:
Yes, the helper function is called `gen_cq_table`. This is already defined.

Vimalakirti commented on src/basic_block/alternating.rs:
```diff
@@ -0,0 +1,162 @@
+// Inputs are X,Y,Z where X alternating with Y equals Z
+// X,Y,Z are multiplied by alpha,beta,alpha_beta to get A,B,C
+impl BasicBlock for AlternatingBasicBlock {
+  fn prove(
```
Clear, I can follow the proving procedure. But we really need a soundness proof for this

arifiorino replied:
I can formalize this, but it will be very similar to the completeness and soundness proofs in the Aurora paper (page 27): https://eprint.iacr.org/2018/828.pdf

Vimalakirti commented on src/basic_block/transpose.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+    r
+  }
+  fn prove(
```
- soundness proof required
- also, maybe we can compare the performance of this block with that of Plonk Permutation later

arifiorino replied:
For the proof, it will be very similar to the completeness and soundness proofs in the Aurora paper (page 27): https://eprint.iacr.org/2018/828.pdf

ddkang replied:
@Vimalakirti please raise an issue or resolve this

Vimalakirti reviewed (approved):
Here is a summary of this review:
- Need to handle shape information later. (the current PR do row-wise provings)
- Some new BasicBlock proving seems to be true (and understandable), but we really need soundness proofs
- Some functions & values duplicate, we can put them in `util.rs` for better maintenance of codes
- Maybe we'll need "verifier_setup" to precompute some msm in verifier's end
- Consider to add formal bench scripts from now on
- It might be good to add examples to showcase how to leverage ML basicblocks (like DivScalar and RoPE)

arifiorino committed:
Remove redundant code

liliatangxy commented on src/basic_block/cq2.rs:
```diff
@@ -0,0 +1,209 @@
+pub struct CQ2BasicBlock {
+  pub table_dict: HashMap<(Fr, Fr), usize>,
+}
+impl BasicBlock for CQ2BasicBlock {
```
How is CQ2 used in the graph and which blocks are using it?

liliatangxy commented on src/basic_block/alternating.rs:
```diff
@@ -0,0 +1,162 @@
+// Inputs are X,Y,Z where X alternating with Y equals Z
+// X,Y,Z are multiplied by alpha,beta,alpha_beta to get A,B,C
+impl BasicBlock for AlternatingBasicBlock {
+  fn prove(
```
Is there a simple way to abstract parts of the aurora prover and verifier? We are currently repeating much of the same proof construction and verifier logic for most of the blocks.

ddkang replied:
@liliatangxy is this resolved? If not, raise an issue

arifiorino replied:
For now this code is only in a few basic blocks so it does not need to be generalized.

liliatangxy reviewed (approved):
I didn't see any issues with the repurposing of aurora for the different blocks. In addition to Bing-Jyue's comments we want to consider
- whether parts of the provers and verifiers that are doing different types of aurora can be partly abstracted, since much of the logic is the same
- whether we can use the same alpha/beta across different blocks to save msms (probably in a different pr)
- sharing the llama graph itself, so that we can clearly see the construction of the graph and how all the blocks are used. The blocks without their own prover/verifier implemented are unclear in how they are being proved

arifiorino committed:
BatchedBasicBlock + Dimension Checks

arifiorino commented:
**New 4/1/2024:**
- Added `BatchedBasicBlock` which runs on `NDArray`
- Switched `Graph` to run on `BatchedBasicBlock`
- `BatchedBasicBlock` checks the dimensions of `model` and `input` before `run`/`setup`/`prove`/`verify` methods

liliatangxy commented on src/basic_block.rs:
```diff
@@ -41,47 +80,155 @@ impl Data {
+    let model = arr_to_vec(model);
+    let model = model.iter().map(|x| x).collect();
+    let inputs = arr_to_vec(inputs);
+    let inputs = inputs.iter().map(|x| x).collect();
```
Is there a reason to use Vec for BasicBlock operations instead of ArrayD?

Vimalakirti replied:
I'm also curious about @liliatangxy 's questions.

ddkang replied:
Is this resolved with the layer issue? @liliatangxy

liliatangxy replied:
As discussed in the meeting I think we can still operate with Vec in BasicBlock but at the layer level we work with ArrayD

liliatangxy commented on src/basic_block.rs:
```diff
@@ -41,47 +80,155 @@ impl Data {
+    assert_eq!(
+      (model.iter().map(|x| x.ndim()).collect(), inputs.iter().map(|x| x.ndim()).collect()),
+      self.basic_block.get_dims()
+    );
```
Is the dimension assertion + ArrayD the only purpose for the `BatchedBasicBlock`? Would it be better to have the dimension check moved into the graph run and the BasicBlocks be changed to use ArrayD?

ddkang replied:
@liliatangxy is this resolved with the layer thing?

liliatangxy replied:
Yes, resolved

liliatangxy reviewed (commented):
Left some initial comments, will look through the rest tonight

Edit: no additional comments

Vimalakirti commented on src/basic_block.rs:
```diff
@@ -122,3 +141,94 @@ pub trait BasicBlock {
+    let model = model.iter().map(|x| x).collect();
+    let inputs = arr_to_vec(inputs);
+    let inputs = inputs.iter().map(|x| x).collect();
+    let outputs = self.basic_block.run(&model, &inputs);
```
Sometimes, we may want to run `self.basic_block.run(&model, &inputs);` N times inside a `BatchedBasicBlock` (as its name suggest(?)) in one ML layer. But currently, we have to create N `BatchedBasicBlock`s to do that. I am not sure if this is a good design choice.

ddkang replied:
Is this the layer issue? If so, raise an issue and resolve this @Vimalakirti

Vimalakirti commented on src/basic_block/add.rs:
```diff
@@ -5,6 +5,9 @@ use rand::rngs::StdRng;
 pub struct AddBasicBlock;
 impl BasicBlock for AddBasicBlock {
+  fn get_dims(&self) -> (Vec<usize>, Vec<usize>) {
+    (vec![], vec![1, 1])
```
We sometimes add model and input together instead of two inputs here. Shall we create an AddModelBasicBlock or  just generalize the current one? (same concerns for other basicblocks like Mul)

ddkang replied:
Make an issue for this

ddkang replied:
@Vimalakirti Did you make an issue for this?

ddkang replied:
If so, resolve this comment

Vimalakirti reviewed (commented):
The main feature added is BatchedBasicBlock, which seems to be an ndarray wrapper for BasicBlock. The additions in other files are mostly fn get_dim() to indicate what the basicblock should take as its input data. 
The summary of this review is as follows:
- (same as Lilia's questions) Do we consider to support ArrayD in BasicBlock? If not, what are the concerns?
- Do we want to generalize BatchedBasicBlock to perform ML layer-level operations in one block?
- Do we want to generalize some BasicBlocks like Add and Mul to support more possible cases during model inference (e.g., model+input instead of input+input)?
Thanks

arifiorino committed:
cleanup
